File: frontend/lib/models/comment.dart
enum CommentDeletedBy { author, admin }

class Comment {
  final int id;
  final String content;
  final int? noticeId; // 💡 공지사항 ID (nullable)
  final int? communityPostId; // 💡 커뮤니티 게시글 ID (nullable)
  final int memberId;
  final String authorName;
  final String createdAt;
  final String? updatedAt;
  final bool isDeleted;
  final String? deletionReason;
  final CommentDeletedBy? deletedBy;
  final int? parentId; // 💡 부모 댓글 ID
  final List<Comment>? children; // 💡 대댓글(자식) 목록

  Comment({
    required this.id,
    required this.content,
    this.noticeId,
    this.communityPostId,
    required this.memberId,
    required this.authorName,
    required this.createdAt,
    this.updatedAt,
    this.isDeleted = false,
    this.deletionReason,
    this.deletedBy,
    this.parentId,
    this.children,
  });

  factory Comment.fromJson(Map<String, dynamic> json) {
    return Comment(
      id: json['id'] ?? 0,
      content: json['content'] ?? '',
      noticeId: json['noticeId'],
      communityPostId: json['communityPostId'],
      memberId: json['memberId'] ?? 0,
      authorName: json['authorName'] ?? '',
      createdAt: json['createdAt'] ?? '',
      updatedAt: json['updatedAt'],
      isDeleted: json['isDeleted'] ?? json['deleted'] ?? false,
      deletionReason: json['deletionReason'],
      deletedBy: _parseDeletedBy(
        json['deletedBy'],
        deletionReason: json['deletionReason'],
      ),
      parentId: json['parentId'],
      children: json['children'] != null
          ? (json['children'] as List).map((i) => Comment.fromJson(i)).toList()
          : null,
    );
  }

  static CommentDeletedBy? _parseDeletedBy(
    Object? value, {
    required Object? deletionReason,
  }) {
    switch (value?.toString().toUpperCase()) {
      case 'AUTHOR':
        return CommentDeletedBy.author;
      case 'ADMIN':
        return CommentDeletedBy.admin;
    }

    // 구형 응답 호환: 관리자 삭제는 사유가 있고, 작성자 삭제는 사유가 없습니다.
    if (deletionReason is String && deletionReason.trim().isNotEmpty) {
      return CommentDeletedBy.admin;
    }
    return null;
  }
}

File: backend/src/main/java/com/ync/ysync/domain/Comment.java
package com.ync.ysync.domain;

import jakarta.persistence.*;
import lombok.AccessLevel;
import lombok.Builder;
import lombok.Getter;
import lombok.NoArgsConstructor;
import org.springframework.data.annotation.CreatedDate;
import org.springframework.data.annotation.LastModifiedDate;
import org.springframework.data.jpa.domain.support.AuditingEntityListener;

import java.time.LocalDateTime;

@Entity
@Getter
@NoArgsConstructor(access = AccessLevel.PROTECTED)
@EntityListeners(AuditingEntityListener.class)
public class Comment {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(columnDefinition = "TEXT", nullable = false)
    private String content;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "notice_id", nullable = true) // 💡 공지사항 댓글일 경우 값이 들어갑니다. (커뮤니티 글일 경우 null)
    private Notice notice;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "community_post_id", nullable = true) // 💡 커뮤니티 게시글 댓글일 경우 값이 들어갑니다. (공지사항 글일 경우 null)
    private CommunityPost communityPost;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "member_id", nullable = false)
    private Member member;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "parent_id", nullable = true)
    private Comment parent;

    @OneToMany(mappedBy = "parent", cascade = CascadeType.ALL, orphanRemoval = true)
    private java.util.List<Comment> children = new java.util.ArrayList<>();

    @CreatedDate
    @Column(updatable = false)
    private LocalDateTime createdAt;

    @LastModifiedDate
    private LocalDateTime updatedAt;

    @Builder
    public Comment(String content, Notice notice, CommunityPost communityPost, Member member, Comment parent) {
        this.content = content;
        this.notice = notice;
        this.communityPost = communityPost;
        this.member = member;
        this.parent = parent;
    }

    // 관리자 삭제 필드
    @Column(nullable = false)
    private boolean isDeleted = false;

    @Column(columnDefinition = "TEXT")
    private String deletionReason;

    public void deleteByAdmin(String reason) {
        this.isDeleted = true;
        this.deletionReason = reason;
    }

    public void restoreByAdmin() {
        this.isDeleted = false;
        this.deletionReason = null;
    }
}
